coderd/prebuilds/state.go

```diff
@@ -0,0 +1,167 @@
+package prebuilds
+
+import (
+	"math"
+	"slices"
+	"time"
+
+	"github.com/coder/quartz"
+
+	"github.com/coder/coder/v2/coderd/database"
+)
+
+func (p PresetState) CalculateActions(clock quartz.Clock, backoffInterval time.Duration) (*ReconciliationActions, error) {
+	// TODO: align workspace states with how we represent them on the FE and the CLI
+	//	     right now there's some slight differences which can lead to additional prebuilds being created
+
+	// TODO: add mechanism to prevent prebuilds being reconciled from being claimable by users; i.e. if a prebuild is
+	// 		 about to be deleted, it should not be deleted if it has been claimed - beware of TOCTOU races!
+
+	var (
+		actual                       int32 // Running prebuilds for active version.
+		desired                      int32 // Active template version's desired instances as defined in preset.
+		eligible                     int32 // Prebuilds which can be claimed.
+		outdated                     int32 // Prebuilds which no longer match the active template version.
+		extraneous                   int32 // Extra running prebuilds for active version (somehow).
+		starting, stopping, deleting int32 // Prebuilds currently being provisioned up or down.
+	)
+
+	if p.Preset.UsingActiveVersion {
+		actual = int32(len(p.Running))
+		desired = p.Preset.DesiredInstances
+	}
+
+	for _, prebuild := range p.Running {
+		if p.Preset.UsingActiveVersion {
+			if prebuild.Ready {
+				eligible++
+			}
+
+			extraneous = int32(math.Max(float64(actual-p.Preset.DesiredInstances), 0))
```

Comment:
- should `extraneous` var be defined outside the loop?
- better to use generic `max` function and avoid conversion to float64?

Reply:
resolved